[#4] session 로그인 구현

ethan-sh-lee opened the pull request:
아이디 비밀번호 확인 후 세션 생성

ethan-sh-lee committed:
session 로그인 구현

아이디 비밀번호 확인 후 세션 생성

f-lab-dev commented on src/main/java/com/project/smartStore/customer/CustomerServiceImpl.java:
```diff
@@ -12,6 +13,24 @@
+      String currentPassword = encryptionConverter.ConvertSHA256WithSalt(loginDTO.getPassword());
+      String storedPassword = getCustomerPassword(loginDTO.getId());
+      if(currentPassword.equals(storedPassword)){
+        session.setAttribute("loginId", loginDTO.getId());
```
여기선 `HttpSession`을 너무 자유롭게 사용하고 있는데 세션을 캡슐화해서 사용하는게 어떨까요?
`SessionLoginService` 같은 것을 만들어서 세션객체를 사용하는 것을 메소드로 제공하면 좋을 것 같아요~

ethan-sh-lee replied:
네 SessionLoginService를 만들어 추가하여 세션 객체를 캡슐화 하겠습니다.

f-lab-dev reviewed (commented):
레디스 세션 저장소도 한번 적용해보시면 좋을 것 같네요~

ethan-sh-lee committed:
sessionLoginService 생성

CustomerServiceImpl에서 SessionLoginService로 session 로그인 관련 관심사를 분리

ethan-sh-lee committed:
sessionLoginService Service애노테이션 추가

누락된 Service 애노테이션 추가

ethan-sh-lee committed:
Redis 종속성 및 설정 추가

Redis 연동 관련 종속성 및 Redis 설정 추가

f-lab-dev commented on src/main/java/com/project/smartStore/customer/CustomerServiceImpl.java:
```diff
@@ -10,8 +13,29 @@
+  private final SessionLoginService sessionLoginService;
+
+  @Override
+  public void loginCustomer(LoginDTO loginDTO, HttpSession session) {
```
`SessionLoginService`가 있음에도 로그인 메소드를 여기에 두신 이유가 있으실까요?

ethan-sh-lee replied:
SessionLoginService을 추가하면서 로그인 로직을 분리하였지만 이제 보니 완전히 분리가 안되었네요. 로그인 로직은 SessionLoginService로 완전히 분리하겠습니다.

f-lab-dev commented on src/main/java/com/project/smartStore/customer/CustomerServiceImpl.java:
```diff
@@ -10,8 +13,29 @@
+      if (currentPassword.equals(storedPassword)) {
+        sessionLoginService.login(session, loginDTO.getId());
+      }
+    } catch (NoSuchAlgorithmException e) {
```
암호화 알고리즘을 찾지 못한게 `loginCustomer`의 책임일까요? 왜 여기서 catch를 해주셨는지 의도가 궁금합니다~

ethan-sh-lee replied:
암호화 알고리즘을 찾는 것은 encryptionConverter의 책임이기 때문에 예외처리 부분을 encryptionConverter로 옮기도록 하겠습니다.

f-lab-dev commented on src/main/java/com/project/smartStore/login/SessionLoginService.java:
```diff
@@ -0,0 +1,15 @@
+  private final String LOGIN_ATTRIBUTE = "LOGIN";
+
+  public void login(HttpSession session, String id) {
+    session.setAttribute(LOGIN_ATTRIBUTE, id);
```
여기선 실제 로그인에 관한 비즈니스로직을 처리하는게 아니라 그냥 세션에 값을 넣어주는 역할만 하고있네요

ethan-sh-lee replied:
로그인 로직이 CustomerService에 남아있네요.. 모든 로그인 로직을 SessionLoginService로 옮기도록 하겠습니다

ethan-sh-lee committed:
리팩토링 : 로그인에 대한 책임을 SessionLoginService로 완전히 분리함

SessionLoginService를 생성하고 로그인에 대한 책음을 분리하였음에도 CustomerServiceImpl에 로그인 로직이 남아있어 이를 완전히 분리하였습니다.
NoSuchAlgorithmException은 CumstomerSeriveImpl의 책임이 아니고 EncryptionConverter의 책임이기 때문에 예외 처리를 EncryptionConverter에서 하도록 변경

f-lab-dev commented on src/main/java/com/project/smartStore/login/SessionLoginService.java:
```diff
@@ -1,15 +1,29 @@
-    session.setAttribute(LOGIN_ATTRIBUTE, id);
+  private final EncryptionConverter encryptionConverter;
+
+  private final CustomerMapper customerMapper;
```
`CustomerMapper`에 직접 의존하기보다는 `CustomerService`를 통해 캡슐화된 메소드를 사용하는게 좋을 것 같습니다~

ethan-sh-lee replied:
CustomService를 통해 CustomMapper를 사용하여 CustomMapper를 캡슐화 하는 방향으로 수정하겠습니다.

f-lab-dev commented on src/main/java/com/project/smartStore/security/EncryptionConverter.java:
```diff
@@ -16,19 +16,24 @@
+      byte[] bytes = messageDigest.digest(inputWithSalt.getBytes(StandardCharsets.UTF_8));
+      return bytesToHex(bytes);
+    } catch (NoSuchAlgorithmException e) {
+      throw new RuntimeException();
```
조금 더 구체적인 예외와 예외메세지, cause까지 추가해주시는게 추후 문제 발생시 디버깅에 도움이 될 것 같습니다~

ethan-sh-lee replied:
네 좀 더 구체적인 커스텀 예외처리를 만들어 수정하겠습니다

ethan-sh-lee committed:
리팩토링 :  SessionLoginService에서 customerMapper 의존을 CustomerServiceImpl으로 옮김

SessionLoginService에서 직접 customerMapper 클래스로 직접 의존하지 않고 유지보수하기 편하게 CustomerService를 통하도록 하여 customerMapper 클래스를 캡슐화함

ethan-sh-lee committed:
기능 : EncryptionConverter RuntimeException에서 세분화된 IllegalAlgorithmException으로 변경

EncryptionConverter에서 Digest로 암호화 알고리즘을 찾는 부분에서 알고리즘을 못찾는 예외가 발생할 경우 기존에는 RuntimeException으로 처리하였지만 좀 더 세분화된 커스텀 예외처리인 IllegalAlgorithmException으로 변경하여 디버깅에 용이하도록 함

f-lab-dev commented on src/main/java/com/project/smartStore/login/SessionLoginService.java:
```diff
@@ -14,12 +14,12 @@
   private final EncryptionConverter encryptionConverter;
 
-  private final CustomerMapper customerMapper;
+  private final CustomerService customerServiceImpl;
```
`customerServiceImpl`이라는 이름은 구체적인 타입에 의존적입니다. 이름을 지을때도 구체적이지 않도록 해주세요~

f-lab-dev commented on src/main/java/com/project/smartStore/security/EncryptionConverter.java:
```diff
@@ -23,7 +23,7 @@ public String ConvertSHA256WithSalt(String inputString) {
       return bytesToHex(bytes);
     } catch (NoSuchAlgorithmException e) {
-      throw new RuntimeException();
+      throw new IllegalAlgorithmException(e.getMessage(), e.getCause());
```
이런식으로 가보면 어떨까요? 왜 이런 방식이 더 나은지도 고민해보시면 좋을 것 같네요

```suggestion
      throw new IllegalAlgorithmException("암호화에 실패하였습니다", e);
```

ethan-sh-lee replied:
예외에 대한 구체적인 메세지를 입력하여 예외 발생시 디버깅이 수월해질 것 같습니다.

f-lab-dev replied:
예외에서 cause란 어떤걸까요? 이걸 넘기면 예외가 발생했을 때 우리 로그에는 어떻게 보일까요?

ethan-sh-lee committed:
스타일 : CustomerServiceimpl 변수를 CustomerService로 변경

ethan-sh-lee committed:
기능 : 예외처리 메세지 수정
